src/meshes/p4est_mesh.jl

```diff
@@ -1021,9 +1136,267 @@ function cubed_sphere_mapping(xi, eta, zeta, inner_radius, thickness, direction)
+            curve_vals[3, p, q] = parse(RealT,current_line[4])
+          end
+        end
+        # Construct the curve interpolant for the current side
+        face_curves[face] = CurvedFaceT(nodes, bary_weights, copy(curve_vals))
+      end
+      # Create the node coordinates on this particular element
+      calc_node_coordinates!(node_coordinates, tree, nodes, face_curves)
+    end
+    # Move file index to the next tree
+    file_idx += 1
+   end
+
+  return file_idx
+end
+
+
+# Given the eight `hex_corners` for a hexahedral element extract
+# the four `face_corners` for a particular `face_index`.
+function get_corners_for_bilinear_interpolant!(face_corners, face_index, hex_corners)
+
+  if face_index == 1
+    @views face_corners[:, 1] .= hex_corners[:, 1]
+    @views face_corners[:, 2] .= hex_corners[:, 2]
+    @views face_corners[:, 3] .= hex_corners[:, 6]
+    @views face_corners[:, 4] .= hex_corners[:, 5]
+  elseif face_index == 2
+    @views face_corners[:, 1] .= hex_corners[:, 4]
+    @views face_corners[:, 2] .= hex_corners[:, 3]
+    @views face_corners[:, 3] .= hex_corners[:, 7]
+    @views face_corners[:, 4] .= hex_corners[:, 8]
+  elseif face_index == 3
+    @views face_corners[:, 1] .= hex_corners[:, 1]
+    @views face_corners[:, 2] .= hex_corners[:, 2]
+    @views face_corners[:, 3] .= hex_corners[:, 3]
+    @views face_corners[:, 4] .= hex_corners[:, 4]
+  elseif face_index == 4
+    @views face_corners[:, 1] .= hex_corners[:, 2]
+    @views face_corners[:, 2] .= hex_corners[:, 3]
+    @views face_corners[:, 3] .= hex_corners[:, 6]
+    @views face_corners[:, 4] .= hex_corners[:, 7]
+  elseif face_index == 5
+    @views face_corners[:, 1] .= hex_corners[:, 5]
+    @views face_corners[:, 2] .= hex_corners[:, 6]
+    @views face_corners[:, 3] .= hex_corners[:, 7]
+    @views face_corners[:, 4] .= hex_corners[:, 8]
+  else # face_index == 6
+    @views face_corners[:, 1] .= hex_corners[:, 1]
+    @views face_corners[:, 2] .= hex_corners[:, 4]
+    @views face_corners[:, 3] .= hex_corners[:, 8]
+    @views face_corners[:, 4] .= hex_corners[:, 5]
+  end
+end
+
+
+# Evaluate a bilinear interpolant at a point (u,v) given the four corners where the face is right-handed
+#      4                3
+#      o----------------o
+#      |                |
+#      |                |
+#      |                |
+#      |                |
+#      |                |
+#      |                |
+#      o----------------o
+#      1                2
+# and return the 3D coordinate point (x, y, z)
+function bilinear_interpolation(face_corners, u, v)
+
+  coordinate = zeros(eltype(face_corners), 3)
+  for j in 1:3
+    coordinate[j] = 0.25 * (  face_corners[j,1] * (1.0 - u) * (1.0 - v)
+                            + face_corners[j,2] * (1.0 + u) * (1.0 - v)
+                            + face_corners[j,3] * (1.0 + u) * (1.0 + v)
+                            + face_corners[j,4] * (1.0 - u) * (1.0 + v) )
+  end
+
+  return coordinate
+end
```

Comment:
Minor comment: You could improve the performance a bit by changing this to `bilinear_interpolation!(coordinate, face_corners, u, v)` where you overwrite `coordinate`. Thus, you can avoid additional allocations of `coordinate = zeros(eltype(face_corners), 3)` for each call. But this function will likely not be a bottleneck, so :shrug:

Reply:
I like it! I was trying to work out how to avoid some of these allocate because I know they are undesired.